Documentation/CONFIG.md
## 🔍 Configuration
You can configure the SDK providing a configuration file with configuration variables or using the configuration method available in the SDK.

### Configuration variables

#### projectToken

* Your project token which can be found in the Exponea APP `Project` -> `Overview`

#### authorization

* Options are `.none` or `.token(token)`.
* Token is an Exponea **public** key. [Configuration guide](./Guide/CONFIGURATION.md) shows where to find it in the Exponea web app.
* For more information, please see [Exponea API documentation](https://docs.exponea.com/reference#access-keys)

#### baseUrl

* If you have custom base URL, you can set up this property.
* Default value `https://api.exponea.com`

#### projectMapping

* In case you need to track events into more than one project, you can define project information for "event types" which should be tracked multiple times.

#### defaultProperties

* A list of properties to be added to all tracking events
* Default value `nil`

#### allowDefaultCustomerProperties

* Flag to apply `defaultProperties` list to `identifyCustomer` tracking event
* Default value `true`

#### automaticSessionTracking

* Flag to control the automatic tracking of `session_start` and `session_end` events.
* Default value `true`

#### sessionTimeout

* Session is the real time spent in the App, it starts when the App is launched and ends when the App goes to background.
* This value will be used to calculate the session timing.
* Default value `6.0` seconds
* The minimum value is `5.0` seconds
* The **recommended** maximum value is `120.0` seconds, but the **absolute** max is `180.0` seconds. More than this, iOS will kill it
* Read more in `Track Events` -> `Track Sessions`

#### automaticPushNotificationTracking - DEPRECATED

* Controls if the SDK will handle push notifications automatically using method swizzling. We decided to deprecate this feature since it uses method swizzling that causes issues when the host application uses multiple SDKs that do the same.

* It is replaced by `pushNotificationTracking`. With `pushNotificationTracking` you're more in control of what's happening inside your app and it also makes debugging easier. When migrating from `automaticPushNotificationTracking` some extra work is required. Please check [Push notifications](./PUSH.md) documentation for more details.
* Default value `true`

#### pushNotificationTracking

* Controls if the SDK will handle push notifications. Registers application to receive push notifications based on `requirePushAuthorization` setting.
* Default value `true`

#### appGroup
* **Required** for the SDK to automatically tracks delivered push notifications. You can find more information in [Push Notifications
](./PUSH.md) documentation.

#### requirePushAuthorization
* SDK can check push notification authorization status([Apple documentation](https://developer.apple.com/documentation/usernotifications/unnotificationsettings/1648391-authorizationstatus)) and only track the push token when the user is authorized to receive push notifications. 
* When disabled, SDK will automatically register for push notifications on app start and track the token to Exponea so your app can receive silent push notifications. 
* When enabled, SDK will automatically register for push notifications if the app is authorized to show push notifications to the user.
* Default value `true`

#### tokenTrackFrequency
	
* Indicates the frequency which the APNS token should be tracked to Exponea
* Default value is `onTokenChange`
* Possible values:
	* `onTokenChange` - tracks push token if differs from previous tracked one
	* `everyLaunch` - tracks push token always
	* `daily` - tracks push token once per day

#### flushEventMaxRetries

* Controls how many times an event should be flushed before aborting. Useful for example if the API is down or some other temporary error happens.
* Default value is `5`.

#### advancedAuthEnabled

* If set, advanced authorization is used for communication with BE for API listed in [JWT Authorization](./AUTHORIZATION.md)
* For more info see [authorization setup](./AUTHORIZATION.md)

#### inAppContentBlocksPlaceholders

* If set, all In-app content blocks will be prefetched right after SDK init

## Configuring the SDK

### 1. Configure SDK programmatically
Since there is a lot of things to configure, configuration is split into few objects that you pass to `Exponea.shared.configure()` function.
``` swift
func configure(
        _ projectSettings: ProjectSettings,
        pushNotificationTracking: PushNotificationTracking,
        automaticSessionTracking: AutomaticSessionTracking = .enabled(),
        defaultProperties: [String: JSONConvertible]? = nil,
        flushingSetup: FlushingSetup = FlushingSetup.default
    )
```

#### ProjectSettings *(required)*
Contains your basic project settings: `projectToken`, `authorization`, `baseUrl` and `projectMapping`.
In most use-cases only `projectToken` and `authorization` is required.

#### PushNotificationTracking *(required)*
Either `.disabled` or `.enabled(appGroup, delegate, requirePushAuthorization, tokenTrackFrequency)`. Only `appGroup` is required for the SDK to function properly.
Setting up `delegate` is the same as settings `Exponea.shared.pushNotificationsDelegate` and is here only for your convenience.

#### automaticSessionTracking
Either `.disabled` or `.enabled(timeout)`. We suggest you use default sessionTimeout and use `.enabled()`, which is also the default value.

#### defaultProperties
As described in configuration variables.

#### flushingSetup
Allows you to setup `flushingMode` and `maxRetries`. By default event flush happens as soon as you track an event(`.immediate`). You can change this behaviour to one of `.manual`, `.automatic`, `periodic(period)`.

#### 💻 Usage
Most common use case:
``` swift
Exponea.shared.configure(
	Exponea.ProjectSettings(
		projectToken: "YOUR PROJECT TOKEN",
		authorization: .token("YOUR ACCESS TOKEN")
	),
	pushNotificationTracking: .enabled(appGroup: "YOUR APP GROUP")
)
```
Disabling all the automatic features of the SDK:
``` swift
Exponea.shared.configure(
	Exponea.ProjectSettings(
		projectToken: "YOUR PROJECT TOKEN",
		authorization: .token("YOUR ACCESS TOKEN")
	),
	pushNotificationTracking: .disabled,
	automaticSessionTracking: .disabled,
	flushingSetup: Exponea.FlushingSetup(mode: .manual)
)
```
Complex use-case:
``` swift
Exponea.shared.configure(
	Exponea.ProjectSettings(
		projectToken: "YOUR PROJECT TOKEN",
		authorization: .token("YOUR ACCESS TOKEN")
		baseUrl: "https://YOUR URL",
		projectMapping: [
			.payment: [
				ExponeaProject(
					baseUrl: "https://YOUR URL",
					projectToken: "YOUR OTHER PROJECT TOKEN",
					authorization: .token("YOUR OTHER ACCESS TOKEN")
				)
			]
		]
	),
	pushNotificationTracking: .enabled(
		appGroup: "YOUR APP GROUP",
		delegate: self,
		requirePushAuthorization: false,
		tokenTrackFrequency: .onTokenChange
	),
	automaticSessionTracking: .enabled(timeout: 123),
	defaultProperties: ["prop-1": "value-1", "prop-2": 123],
	flushingSetup: Exponea.FlushingSetup(mode: .periodic(100), maxRetries: 5),
	advancedAuthEnabled: true
)
```


### 2. Using a configuration file - LEGACY
> Configuring the SDK programmatically is preferred. SDK still supports configuring the SDK using a plist file for backwards compatibility.

Create a configuration `.plist` file containing configuration variables. You need to specify all the required ones. For the rest, default values will be used.
``` swift
public func configure(plistName: String)
```

#### 💻 Usage

*ExponeaConfig.plist*

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<plist version="1.0">
<dict>
	<key>projectToken</key>
	<string>testToken</string>
	<key>sessionTimeout</key>
	<integer>20</integer>
	<key>projectMapping</key>
	<dict>
		<key>INSTALL</key>
		<array>
			<dict>
				<key>projectToken</key>
				<string>testToken1</string>
				<key>authorization</key>
				<string>Token authToken1</string>
			</dict>
		</array>
		<key>TRACK_EVENT</key>
		<array>
			<dict>
				<key>projectToken</key>
				<string>testToken2</string>
				<key>authorization</key>
				<string>Token authToken2</string>
			</dict>
			<dict>
				<key>projectToken</key>
				<string>testToken3</string>
				<key>authorization</key>
				<string></string>
			</dict>
		</array>
		<key>PAYMENT</key>
		<array>
			<dict>
				<key>baseUrl</key>
				<string>https://mock-base-url.com</string>
				<key>projectToken</key>
				<string>testToken4</string>
				<key>authorization</key>
				<string>Token authToken4</string>
			</dict>
		</array>
	</dict>
	<key>autoSessionTracking</key>
	<false/>
</dict>
</plist>
```

```
Exponea.shared.configure(plistName: "ExponeaConfig.plist")
```

### AppInbox detail image inset

The inset is by default 56 for normal purpose (small title, without searchBar bellow etc). If you need to change its value, you can set value whatewer you want. 0 is without space bewteen status bar and image - for transaprent navigationBar for example.


Documentation/RELEASE_NOTES.md
## Release Notes
## Release Notes for 2.19.0
#### October 27, 2023
* Features
  * Xcode 15 and iOS 17 support added
    * iOS minimal version increased to 12
  * Push notification token tracking frequency has been documented
  * Anonymize feature has been described with more details in documentation
  * Rich Push notification has clickable image with default action
  * Viewport meta tag was removed from forbidden constructs for all HTML messages due to scaling issues
  * Debug mode was refactored
* Bug Fixes
  * Fixed: Action button not working for HTML In-app messages created via Visual editor
  * Fixed: Push notification payload docu has wrong field names
  * Fixed: Multiple flush timers invocation


## Release Notes for 2.18.0
#### September 14, 2023
* Features
  * Added EXP prefix for for SDK-related log output
  * Added dynamic scaling for App Inbox PUSH message
  * Add support for darkmode in App Inbox PUSH messages and native In-app messages in by isDarkMode
  * Documentation extension with PUSH notification payload structure description, PUSH handling and more
  * Small internal refactoring for better compatibility with MAUI wrapper
* Bug Fixes
  * Fixed: Image is under navbar in App Inbox PUSH message detail


## Release Notes for 2.17.0
#### August 02, 2023
* Features
  * In-app content block feature has been added into SDK
  * Handling of nil/null values inside objects sending tracking in requestFactory to avoid crashes
  * Documentation is updated to describe fetching of In-app messages while identifyCustomer process in detail
  * New devices supported for tracking events
* Bug Fixes
  * Fixed: Action click event from In-app messages, and App Inbox HTML Inbox messages tracked button text with HTML tags
  * Fixed: banner_type for HTML In-app messages is null instead of freeform
  * Fixed: Webview for HTML In-app messages and App Inbox HTML Inbox messages not fully offline

## Release Notes for 2.16.4
#### July 11, 2023
* Features
  * XCode 14.3 support added
* Bug Fixes
  * Fixed: Property `platform` was not tracked for campaign events


## Release Notes for 2.16.3
#### June 01, 2023
* Features
  * Brings public visibility to App Inbox style API 


## Release Notes for 2.16.2
#### May 22, 2023
* Features
  * New AppInbox provider has been created for easier styling of UI components
  * API for push token handling as string and push notification tracking
  * Small internal refactoring for better compatibility with other wrappers
  * Documentation about FCMID handling was described more deeply


## Release Notes for 2.16.1
#### May 04, 2023
* Bug Fixes
  * Fixed: Carthage dependency linking and wrong file name


## Release Notes for 2.16.0
#### April 25, 2023
* Features
  * Ability to track a user 'interaction' while closing InApp message
  * Support section added to main Readme
  * InApp loading flow has been described more deeply in documentation
  * Push notification track action has been described more deeply in documentation
  * SDK initialization performance improve
* Bug Fixes
  * Fixed: SwiftSoup takes large space because of xcframework assignment
  * Fixed: Marking of App Inbox message as read failed due to invalid usage of Customer IDs
  * Fixed: Empty in-app message response did not clear the cache so previously loaded messages persisted there
  * Fixed: Remove invalid link to developers hub from Useful links


## Release Notes for 2.15.2
#### March 06, 2023
* Bug Fixes
  * Fixed: AppInbox detail layout fixes


## Release Notes for 2.15.1
#### February 27, 2023
* Bug Fixes
  * Fixed: Fixed: App Inbox layout was reworked from using storyboard to native code implementation for better compatibility with SDK wrappers


## Release Notes for 2.15.0
#### February 16, 2023
* Features
  * Extended App Inbox functionality with new HTML message type - Inbox message
  * All SDK methods are invoked after SDK has been initialised
  * Added support for Customer token authorization

## Release Notes for 2.14.0
#### December 15, 2022
* Features
  * Added App Inbox feature with PUSH message type support

## Release Notes for 2.13.1
#### November 24, 2022
* Features
  * Added Configuration flag to be able to disable tracking of default properties along with customer properties
  * Increased support for iOS SDK 11+
  * Updated guide documentation for Push notifications service setup
* Bug Fixes
  * Fixed: Handling of InApp messages may cause race-condition crash
  * Fixed: Handling of universal links in HTML InApp messages was not working properly


## Release Notes for 2.13.0
#### October 19, 2022
* Features
  * Added feature to handle tracking consent according to DSGVO/GDPR
  * Guiding documentation added for Push notification update after certain circumstances
* Bug Fixes
  * Fixed: Webview configuration updated according to newer XCode14


## Release Notes for 2.12.3
#### September 20, 2022
* Bug Fixes
  * Fixed: Webview configuration setup compatible with swift5.7


## Release Notes for 2.12.2
#### September 06, 2022
* Bug Fixes
  * Fixed: DatabaseManager init with Bundle creation for Swift Package Manager usage


## Release Notes for 2.12.1
#### September 01, 2022
* Bug Fixes
  * Fixed: License established to MIT
  * Fixed: Warn log for developer for old SDK usage has been fixed for compatibility with SDK wrappers


## Release Notes for 2.12.0
#### August 09, 2022
* Features

ExponeaSDK/ExponeaSDKShared/Classes/Configuration.swift
//
//  Configuration.swift
//  ExponeaSDK
//
//  Created by Ricardo Tokashiki on 06/04/2018.
//  Copyright © 2018 Exponea. All rights reserved.
//

import Foundation
import CoreData

/// A configuration object used to configure Exponea when initialising.
public struct Configuration: Codable, Equatable {
    public internal(set) var projectMapping: [EventType: [ExponeaProject]]?
    public internal(set) var projectToken: String
    public internal(set) var authorization: Authorization = .none
    public internal(set) var baseUrl: String = Constants.Repository.baseUrl
    public var inAppContentBlocksPlaceholders: [String]?
    public var defaultProperties: [String: JSONConvertible]?
    public var sessionTimeout: Double = Constants.Session.defaultTimeout
    public var automaticSessionTracking: Bool = true

    /// If enabled, will swizzle default push notifications methods and functions and automatically
    /// listen to updates for tokens or push settings.
    public internal(set) var automaticPushNotificationTracking: Bool = true

    /// If true, push notification registration and push token tracking is only done if the device is authorized
    /// to receive push notifications.
    /// Disabling is useful for silent push notifications that don't require authorization.
    /// In that case, registration and token tracking is done at app start automatically.
    public internal(set) var requirePushAuthorization: Bool = true

    /// If automatic push notification tracking is enabled, this can be used to determine how often
    /// should the push notification token be sent to Exponea.
    public internal(set) var tokenTrackFrequency: TokenTrackFrequency = .onTokenChange

    /// App group is used when push notification data is shared among service or content extensions.
    /// This is required for tracking delivered push notifications properly.
    public internal(set) var appGroup: String?

    /// The maximum amount of retries before a flush event is considered as invalid and deleted from the database.
    public internal(set) var flushEventMaxRetries: Int = Constants.Session.maxRetries

    /// If true, default properties are applied also for 'identifyCustomer' event.
    public internal(set) var allowDefaultCustomerProperties: Bool = true

    /// If true, advanced authorization is used for communication with BE
    public var advancedAuthEnabled: Bool = false

    /// Advanced authorization provider instance
    public internal(set) var customAuthProvider: AuthorizationProviderType?

    /// Is dark mode enabled
    public internal(set) var isDarkModeEnabled: Bool?

    ///  App inbox detail image inset
    public internal(set) var appInboxDetailImageInset: CGFloat?

    enum CodingKeys: String, CodingKey {
        case projectMapping
        case projectToken
        case authorization
        case baseUrl
        case defaultProperties
        case sessionTimeout
        case automaticSessionTracking
        case automaticPushNotificationTracking
        case requirePushAuthorization
        case tokenTrackFrequency
        case appGroup
        case flushEventMaxRetries
        case allowDefaultCustomerProperties
        case advancedAuthEnabled
        case isDarkModeEnabled
        case appInboxDetailImageInset
    }

    /// Creates the configuration object with the provided properties.
    ///
    /// - Parameters:
    ///   - projectToken: The project token used for connecting with Exponea.
    ///   - projectMapping: Optional project mapping if you wish to send events to different projects.
    ///   - authorization: The authorization you want to use when tracking events.
    ///   - baseUrl: Your API base URL that the SDK will connect to.
    ///   - defaultProperties: Custom properties to be tracked in every event.
    ///   - advancedAuthEnabled: Flag if advanced authorization used for communication with BE
    public init(projectToken: String?,
                projectMapping: [EventType: [ExponeaProject]]? = nil,
                authorization: Authorization,
                baseUrl: String?,
                appGroup: String? = nil,
                defaultProperties: [String: JSONConvertible]? = nil,
                inAppContentBlocksPlaceholders: [String]? = nil,
                allowDefaultCustomerProperties: Bool? = nil,
                advancedAuthEnabled: Bool? = nil,
                isDarkModeEnabled: Bool? = nil,
                appInboxDetailImageInset: CGFloat? = 56
    ) throws {
        guard let projectToken = projectToken else {
            throw ExponeaError.configurationError("No project token provided.")
        }
        self.projectToken = projectToken
        self.projectMapping = projectMapping
        self.authorization = authorization
        self.appGroup = appGroup
        self.defaultProperties = defaultProperties
        self.allowDefaultCustomerProperties = allowDefaultCustomerProperties ?? true
        self.advancedAuthEnabled = advancedAuthEnabled ?? false
        self.inAppContentBlocksPlaceholders = inAppContentBlocksPlaceholders
        self.appInboxDetailImageInset = appInboxDetailImageInset
        if let url = baseUrl {
            self.baseUrl = url
        }
        if (self.advancedAuthEnabled) {
            self.customAuthProvider = try loadCustomAuthProvider()
        }
        try self.validate()
        self.isDarkModeEnabled = isDarkModeEnabled ?? false
    }

    public init(
        projectToken: String,
        projectMapping: [EventType: [ExponeaProject]]?,
        authorization: Authorization = .none,
        baseUrl: String,
        defaultProperties: [String: JSONConvertible]?,
        inAppContentBlocksPlaceholders: [String]? = nil,
        sessionTimeout: Double,
        automaticSessionTracking: Bool = true,
        automaticPushNotificationTracking: Bool,
        requirePushAuthorization: Bool = true,
        tokenTrackFrequency: TokenTrackFrequency,
        appGroup: String?,
        flushEventMaxRetries: Int,
        allowDefaultCustomerProperties: Bool?,
        advancedAuthEnabled: Bool?,
        isDarkModeEnabled: Bool? = nil,
        appInboxDetailImageInset: CGFloat? = 56
    ) throws {
        self.projectToken = projectToken
        self.projectMapping = projectMapping
        self.authorization = authorization
        self.baseUrl = baseUrl
        self.defaultProperties = defaultProperties
        self.inAppContentBlocksPlaceholders = inAppContentBlocksPlaceholders
        self.sessionTimeout = sessionTimeout
        self.automaticSessionTracking = automaticSessionTracking
        self.automaticPushNotificationTracking = automaticPushNotificationTracking
        self.requirePushAuthorization = requirePushAuthorization
        self.tokenTrackFrequency = tokenTrackFrequency
        self.appGroup = appGroup
        self.flushEventMaxRetries = flushEventMaxRetries
        self.allowDefaultCustomerProperties = allowDefaultCustomerProperties ?? true
        self.advancedAuthEnabled = advancedAuthEnabled ?? false
        self.appInboxDetailImageInset = appInboxDetailImageInset
        if (self.advancedAuthEnabled) {
            self.customAuthProvider = try loadCustomAuthProvider()
        }
        try self.validate()
        self.isDarkModeEnabled = isDarkModeEnabled ?? false
    }

    /// Creates the Configuration object from a plist file.
    ///
    /// - Parameter plistName: The name of the plist file you want to load configuration from.
    public init(plistName: String) throws {
        for bundle in Bundle.allBundles {
            let fileName = plistName.replacingOccurrences(of: ".plist", with: "")
            guard let fileURL = bundle.url(forResource: fileName, withExtension: "plist") else {
                continue
            }

            Exponea.logger.log(.verbose, message: """
                Found configuration file with name \(fileName) in bundle: \(bundle.bundlePath)
                """)

            // Load the data
            let data = try Data(contentsOf: fileURL)

            // Decode from plist
            self = try PropertyListDecoder().decode(Configuration.self, from: data)
            if (self.advancedAuthEnabled) {
                self.customAuthProvider = try loadCustomAuthProvider()
            }
            try self.validate()

            // Stop if we found the file and decoded successfully
            return
        }
        throw ExponeaError.configurationError("Configuration plist not found.")
    }

    // MARK: - Decodable -

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        // Project token
        guard let projectToken = try container.decodeIfPresent(String.self, forKey: .projectToken) else {
            throw ExponeaError.configurationError("No project token provided.")
        }
        self.projectToken = projectToken

        // Base URL
        if let baseUrl = try container.decodeIfPresent(String.self, forKey: .baseUrl) {
            self.baseUrl = baseUrl
        }

        // Authorization
        if let authorization = try container.decodeIfPresent(Authorization.self, forKey: .authorization) {
            self.authorization = authorization
        }

        // Project token mapping
        if let dictionary = try container.decodeIfPresent(
            Dictionary<String, [ExponeaProject]>.self, forKey: .projectMapping) {
            var mapping: [EventType: [ExponeaProject]] = [:]
            for (_, element: (key: event, value: projectArray)) in dictionary.enumerated() {
                guard let eventType = EventType(rawValue: event) else { continue }
                mapping[eventType] = projectArray
            }
            self.projectMapping = mapping
        }

        // Session timeout
        if let sessionTimeout = try container.decodeIfPresent(Double.self, forKey: .sessionTimeout) {
            self.sessionTimeout = sessionTimeout
        }

        // Automatic sessiont racking
        if let automaticSessionTracking = try container.decodeIfPresent(
            Bool.self, forKey: .automaticSessionTracking) {
            self.automaticSessionTracking = automaticSessionTracking
        }

        // Automatic push notifications
        if let automaticPushNotificationTracking = try container.decodeIfPresent(
            Bool.self, forKey: .automaticPushNotificationTracking) {
            self.automaticPushNotificationTracking = automaticPushNotificationTracking
        }

        // Requiring push authorization
        if let requirePushAuthorization = try container.decodeIfPresent(
            Bool.self, forKey: .requirePushAuthorization) {
            self.requirePushAuthorization = requirePushAuthorization
        }

        // Token track frequency
        if let tokenTrackFrequency = try container.decodeIfPresent(
            TokenTrackFrequency.self, forKey: .tokenTrackFrequency) {
            self.tokenTrackFrequency = tokenTrackFrequency
        }

        // Flush event max retries
        if let flushEventMaxRetries = try container.decodeIfPresent(
            Int.self, forKey: .flushEventMaxRetries) {
            self.flushEventMaxRetries = flushEventMaxRetries
        }
        
        // isDarkModeEnabled
        if let isDarkModeEnabled = try container.decodeIfPresent(
            Bool.self, forKey: .isDarkModeEnabled) {
            self.isDarkModeEnabled = isDarkModeEnabled
        }
        
        // appInboxDetailImageInset
        if let appInboxDetailImageInset = try container.decodeIfPresent(
            CGFloat.self, forKey: .appInboxDetailImageInset) {
            self.appInboxDetailImageInset = appInboxDetailImageInset
        }

        // App group
        if let appGroup = try container.decodeIfPresent(String.self, forKey: .appGroup) {
            self.appGroup = appGroup
        }

        // Default properties
        if let defaultDictionary = try container.decodeIfPresent(
            [String: JSONValue].self, forKey: .defaultProperties) {
            var properties: [String: JSONConvertible] = [:]
            defaultDictionary.forEach({ property in
                properties[property.key] = property.value.jsonConvertible
            })
            guard !properties.isEmpty else { return }
            self.defaultProperties = properties
        }

        // Default properties usage for Identify Customer event
        if let allowDefaultCustomerProperties = try container.decodeIfPresent(
            Bool.self, forKey: .allowDefaultCustomerProperties) {
            self.allowDefaultCustomerProperties = allowDefaultCustomerProperties
        } else {
            self.allowDefaultCustomerProperties = true
        }

        // Advanced auth token
        if let advancedAuthEnabled = try container.decodeIfPresent(
            Bool.self, forKey: .advancedAuthEnabled) {
            self.advancedAuthEnabled = advancedAuthEnabled
        } else {
            self.advancedAuthEnabled = false
        }

        // Advanced auth provider
        if self.advancedAuthEnabled {
            self.customAuthProvider = try loadCustomAuthProvider()
        }
        try self.validate()
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        if let mapping = projectMapping {
            let projectMappingWithStringKeys = Dictionary(
                uniqueKeysWithValues: mapping.map { (key: EventType, value: [ExponeaProject]) in (key.rawValue, value) }
            )
            try container.encode(projectMappingWithStringKeys, forKey: .projectMapping)
        }
        try container.encode(projectToken, forKey: .projectToken)
        try container.encode(authorization, forKey: .authorization)
        try container.encode(baseUrl, forKey: .baseUrl)
        try container.encode(defaultProperties?.mapValues { $0.jsonValue }, forKey: .defaultProperties)
        try container.encode(sessionTimeout, forKey: .sessionTimeout)
        try container.encode(automaticSessionTracking, forKey: .automaticSessionTracking)
        try container.encode(automaticPushNotificationTracking, forKey: .automaticPushNotificationTracking)
        try container.encode(requirePushAuthorization, forKey: .requirePushAuthorization)
        try container.encode(tokenTrackFrequency, forKey: .tokenTrackFrequency)
        try container.encode(appGroup, forKey: .appGroup)
        try container.encode(flushEventMaxRetries, forKey: .flushEventMaxRetries)
        try container.encode(allowDefaultCustomerProperties, forKey: .allowDefaultCustomerProperties)
        try container.encode(advancedAuthEnabled, forKey: .advancedAuthEnabled)
    }

    public static func == (lhs: Configuration, rhs: Configuration) -> Bool {
        return
            lhs.projectMapping == rhs.projectMapping &&
            lhs.projectToken == rhs.projectToken &&
            lhs.authorization == rhs.authorization &&
            lhs.baseUrl == rhs.baseUrl &&
            lhs.defaultProperties?.mapValues { $0.jsonValue } == rhs.defaultProperties?.mapValues { $0.jsonValue } &&
            lhs.sessionTimeout == rhs.sessionTimeout &&
            lhs.automaticSessionTracking == rhs.automaticSessionTracking &&
            lhs.automaticPushNotificationTracking == rhs.automaticPushNotificationTracking &&
            lhs.requirePushAuthorization == rhs.requirePushAuthorization &&
            lhs.tokenTrackFrequency == rhs.tokenTrackFrequency &&
            lhs.appGroup == rhs.appGroup &&
            lhs.flushEventMaxRetries == rhs.flushEventMaxRetries &&
            lhs.allowDefaultCustomerProperties == rhs.allowDefaultCustomerProperties &&
            lhs.advancedAuthEnabled == rhs.advancedAuthEnabled
    }
}

extension Configuration {
    public func projects(for eventType: EventType) -> [ExponeaProject] {
        var projects: [ExponeaProject] = [mainProject]
        if let mapping = projectMapping, let mappedTokens = mapping[eventType] {
            projects.append(contentsOf: mappedTokens)
        }
        return projects
    }

    public var mainProject: ExponeaProject {
        ExponeaProject(baseUrl: baseUrl, projectToken: projectToken, authorization: authorization)
    }

    /// Returns ExponeaProject that:
    /// - contains CustomerID auth token if AuthProvider is registered
    /// - contains Api auth token otherwise
    /// !!! Access it in background thread due to possibility of fetching of Customer Token value
    public var mutualExponeaProject: ExponeaProject {
        guard let customAuthProvider = self.customAuthProvider else {
            return mainProject
        }
        let authToken = customAuthProvider.getAuthorizationToken()
        let authorization: Authorization
        if let authToken = authToken, !authToken.isEmpty {
            authorization = .bearer(token: authToken)
        } else {
            authorization = .none
        }
        return ExponeaProject(
            baseUrl: baseUrl,
            projectToken: projectToken,
            authorization: authorization
        )
    }

    private func loadCustomAuthProvider() throws -> AuthorizationProviderType? {
        let className = "ExponeaAuthProvider"
        guard let foundClass = NSClassFromString(className) else {
            // valid exit
            return nil
        }
        guard let asNSObjectClass = foundClass as? NSObject.Type else {
            Exponea.logger.log(.error, message: "Class '\(className)' does not conform to NSObject")
            throw ConfigurationValidationError.advancedAuthInvalid(
                "Class '\(className)' does not conform to NSObject"
            )
        }
        guard let asProviderClass = asNSObjectClass as? AuthorizationProviderType.Type else {
            Exponea.logger.log(.error, message: "Class '\(className)' does not conform to AuthorizationProviderType")
            throw ConfigurationValidationError.advancedAuthInvalid(
                "Class '\(className)' does not conform to AuthorizationProviderType"
            )
        }
        return asProviderClass.init()
    }
}

extension Configuration: CustomStringConvertible {
    public var description: String {
        var text = "[Configuration]\n"

        if let mapping = projectMapping {
            text += "Project Token Mapping: \(mapping)\n"
        }

        text += "Project Token: \(projectToken)\n"

        if let defaultProperties = defaultProperties {
            text += "Default Attributes: \(defaultProperties)\n"
        }

        text += """
        Authorization: \(authorization)
        Base URL: \(baseUrl)
        Session Timeout: \(sessionTimeout)
        Automatic Session Tracking: \(automaticSessionTracking)
        Automatic Push Notification Tracking: \(automaticPushNotificationTracking)
        Require Push Authorization: \(requirePushAuthorization)
        Token Track Frequency: \(tokenTrackFrequency)
        Flush Event Max Retries: \(flushEventMaxRetries)
        App Group: \(appGroup ?? "not configured")
        Default Customer Props allowed: \(allowDefaultCustomerProperties)
        Advanced authorization Enabled: \(advancedAuthEnabled)
        """

        return text
    }

    /// Returns the hostname based on the baseUrl value.
    public var hostname: String {
        guard let components = URLComponents(string: baseUrl),
            let host = components.host else {
            Exponea.logger.log(.warning, message: "Can't get URL components from baseUrl, check your baseUrl.")
            return baseUrl
        }

        return host
    }
}

@objc(AuthorizationProviderType)
public protocol AuthorizationProviderType {
    init()
    func getAuthorizationToken() -> String?
}
